src/modules/rest/extjs/src/main/java/it/geosolutions/geostore/services/rest/impl/RESTExtJsServiceImpl.java

```diff
@@ -353,29 +385,47 @@ public ExtResourceList getExtResourcesList(
     }
 
     /**
-     * This internal method allows us to filter out "unadvertised" resources for
-     * non-admin/non-owners
+     * Filters out unadvertised resources and adds permission information to the resources.
      *
-     * @param resources
-     * @param authUser
+     * @param foundResources
+     * @param user
      * @return
      */
-    private List<Resource> filterOutUnadvertisedResources(List<Resource> resources, User authUser) {
-        List<Resource> filteredList = new ArrayList<>();
-        for (Resource r : resources) {
-            // get security rules for user
-            User owner = null;
-            for (SecurityRule rule :
-                    resourceService.getUserSecurityRule(authUser.getName(), r.getId())) {
-                owner = rule.getUser();
-            }
-            if (r.isAdvertised()
-                    || (authUser.getRole().equals(Role.ADMIN)
-                            || (owner != null && owner.getId().equals(authUser.getId())))) {
-                filteredList.add(r);
-            }
-        }
-        return filteredList;
+    private List<Resource> handleFoundResources(List<Resource> foundResources, User user) {
+        return foundResources.stream()
+                .filter(r -> isResourceAvailableForUser(r, user))
+                .map(r -> addPermissionsToResource(r, user))
+                .collect(Collectors.toList());
+    }
+
+    /**
+     * This internal method allows us to know if we filter out "unadvertised" resources for
+     * non-admin/non-owners, keeping only owned resources.
+     *
+     * @param resource
+     * @param user
+     * @return <code>true</code> if the resource should be visible to the user, <code>false</code>
+     *     otherwise
+     */
+    private boolean isResourceAvailableForUser(Resource resource, User user) {
+        return resource.isAdvertised()
+                || user.getRole().equals(Role.ADMIN)
+                || isUserOwner(user, resource);
+    }
+
+    private boolean isUserOwner(User authUser, Resource r) {
+        return resourceService.getSecurityRules(r.getId()).stream()
+                .map(SecurityRule::getUser)
+                .filter(Objects::nonNull)
+                .anyMatch(ruleUser -> ruleUser.getId().equals(authUser.getId()));
+    }
+
+    private Resource addPermissionsToResource(Resource resource, User user) {
+        ResourceEnvelop envelope = new ResourceEnvelop(resource, user);
```

Comment:
This internal class encapsulates a chunk of permission logic. Some logic duplicates the code in ResourceServiceImpl and RESTExtJsServiceImpl.
If ResourceEnvelop is primarily for permission decoration, consider a single approach: either do it in the service layer, or do it in a dedicated “decorator” layer. Minimizing overlap will make it easier to maintain.

Reply:
Made the `ResourceEnvelop` inner class to use the extracted `PermissionService`.